In [1]:
import numpy as np
import random
import sys
sys.path.insert(0, '/Users/trimcao/Dropbox/Richardson/Fall-2017/cs6375-ml-ruozzi/solution/lib')
sys.path.insert(0, '/home/trimcao/Dropbox/Richardson/Fall-2017/cs6375-ml-ruozzi/solution/lib')
from DecisionTree import DecisionTree, BoostedTree

### Read in data

In [2]:
def read_in(file_path):
    X = []
    y = []
    with open(file_path, 'r') as f:
        for line in f:
            info = line.strip('\n').split(',')
            X.append([int(i) for i in info[1:]])
            y.append(int(info[0]))
    X = np.array(X)
    y = np.array(y)
    # make the labels -1 and 1
    y[y==0] = -1
    return X, y

In [3]:
X, y = read_in('hw3_data/heart_train.data')

In [4]:
X_test, y_test = read_in('hw3_data/heart_test.data')

In [5]:
N = X.shape[0]
weights = np.array([1/N for i in range(N)])

In [6]:
print(X[0])

[0 0 0 1 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 0 0 0]


In [7]:
boost = BoostedTree(X, y)

In [12]:
trees = boost.gen_tree5([10, 12, 14], X, y)

In [13]:
print(len(trees))
for each in trees:
    preds = each.predict(X)
    print(np.mean(preds==y))

16
0.5
0.5
0.5625
0.5625
0.5625
0.5625
0.625
0.625
0.375
0.375
0.4375
0.4375
0.4375
0.4375
0.5
0.5


### Test adaBoost

In [8]:
boost = BoostedTree(X, y, M=3)
boost.fit_hw(X, y, X_test, y_test)

Iteration: 1
Accuracy on train set: 0.8125
Accuracy on test set: 0.75935828877
Iteration: 2
Accuracy on train set: 0.8125
Accuracy on test set: 0.75935828877
Iteration: 3
Accuracy on train set: 0.8125
Accuracy on test set: 0.818181818182


In [8]:
preds = boost.predict(X)
print(np.mean(preds==y))

0.8125


In [9]:
test_preds = boost.predict(X_test)
print(np.mean(test_preds==y_test))

0.818181818182


In [10]:
print(boost.stages)

[0.73316853439671348, 0.49926441505556374, 0.33086549216328331]


In [11]:
tree1 = boost.trees[0].tree
print(tree1.split_feature)
print(tree1.children[0].split_feature)
print(tree1.children[1].split_feature)
print(tree1.children[1].predict)
left = tree1.children[0]
print(left.children[0].split_feature)
print(left.children[1].split_feature)
print(left.children[1].predict)
left2 = left.children[0]
print(left2.children[0].predict)
print(left2.children[1].predict)

7
10
None
1
15
None
1
-1
1


In [12]:
tree = boost.trees[1].tree
print(tree.split_feature)
print(tree.children[0].split_feature)
print(tree.children[1].split_feature)
print(tree.children[1].predict)
left = tree.children[0]
print(left.children[0].split_feature)
print(left.children[1].split_feature)
print(left.children[1].predict)
left2 = left.children[0]
print(left2.children[0].predict)
print(left2.children[1].predict)

11
19
None
1
21
None
1
-1
1


In [13]:
tree = boost.trees[2].tree
print(tree.split_feature)
print(tree.children[0].split_feature)
print(tree.children[1].split_feature)
print(tree.children[1].predict)
left = tree.children[0]
print(left.children[0].split_feature)
print(left.children[1].split_feature)
print(left.children[1].predict)
left2 = left.children[0]
print(left2.children[0].predict)
print(left2.children[1].predict)

12
2
None
1
21
None
-1
1
-1


### adaBoost with M=10

In [14]:
boost = BoostedTree(X, y, M=10)
boost.fit_hw(X, y, X_test, y_test)

Iteration: 1
Accuracy on train set: 0.8125
Accuracy on test set: 0.75935828877
Iteration: 2
Accuracy on train set: 0.8125
Accuracy on test set: 0.75935828877
Iteration: 3
Accuracy on train set: 0.8125
Accuracy on test set: 0.818181818182
Iteration: 4
Accuracy on train set: 0.875
Accuracy on test set: 0.791443850267
Iteration: 5
Accuracy on train set: 0.8
Accuracy on test set: 0.818181818182
Iteration: 6
Accuracy on train set: 0.9
Accuracy on test set: 0.75935828877
Iteration: 7
Accuracy on train set: 0.9
Accuracy on test set: 0.75935828877
Iteration: 8
Accuracy on train set: 0.9125
Accuracy on test set: 0.754010695187
Iteration: 9
Accuracy on train set: 0.925
Accuracy on test set: 0.732620320856
Iteration: 10
Accuracy on train set: 0.9125
Accuracy on test set: 0.743315508021


In [15]:
print(boost.errors)

[0.18750000000000003, 0.26923076923076916, 0.3403508771929824, 0.33719604863221897, 0.36971122376230831, 0.33136955688368658, 0.34394375957211726, 0.3472816724785559, 0.37868432819030851, 0.37269850036829033]


### Train a decision tree with unlimited depth

In [8]:
clf = DecisionTree()
clf.fit(X, y, weights=weights)

In [9]:
print(clf.feature_dict)

{0: {0, 1}, 1: {0, 1}, 2: {0, 1}, 3: {0, 1}, 4: {0, 1}, 5: {0, 1}, 6: {0, 1}, 7: {0, 1}, 8: {0, 1}, 9: {0, 1}, 10: {0, 1}, 11: {0, 1}, 12: {0, 1}, 13: {0, 1}, 14: {0, 1}, 15: {0, 1}, 16: {0, 1}, 17: {0, 1}, 18: {0, 1}, 19: {0, 1}, 20: {0, 1}, 21: {0, 1}}


In [10]:
print(clf.predict_single(X_test[2]))

-1


In [11]:
preds = clf.predict(X_test)

In [12]:
print(np.mean(preds==y_test))

0.743315508021


In [13]:
print(clf.depth)

22


In [14]:
train_preds = clf.predict(X)
print(np.mean(train_preds==y))

0.9375


Depth = 22 means overfit.

### Train a decision tree with depth = 2

In [15]:
clf = DecisionTree(depth=2)
clf.fit(X, y, weights)

In [16]:
preds2 = clf.predict(X_test)

In [17]:
print(np.mean(preds2==y_test))

0.711229946524


Accuracy is not bad at all (0.711229946524).

### Probability with weights

In [18]:
weights = np.array([0.2, 0.5, 0.2, 0.1])
y_small = np.array([1, 0, 1, 0])
x_small = np.array([1, 1, 0, 0])